include/pybind11/pytypes.h

```diff
@@ -298,6 +298,22 @@ class handle : public detail::object_api<handle> {
 
 #ifdef PYBIND11_HANDLE_REF_DEBUG
 private:
+    void throw_gilstate_error(const std::string &function_name) const {
+        fprintf(stderr,
+                "%s is being called while PyGILState_Check() is failing. "
+                "This usually indicates that you are calling pybind11 functions without holding "
+                "the GIL or before "
+                "Python (and/or this Python module) is finished initializing.\n",
+                function_name.c_str());
+        if (Py_TYPE(m_ptr)->tp_name != nullptr) {
```

Comment:
I don't think your logic in obj_class_name is actually correct? I don't think it would handle metaclasses right.

Reply:
Intrigued!

Nutshell background: I didn't invent this, I copied it over [from PyCLIF](https://github.com/google/clif/blob/db5be2f549793c9d641eae027056f419a41c70ce/clif/python/runtime.cc#L159). I'm guessing the original author(s) didn't consider metaclasses, and neither did I.

How do we make this right for metaclasses? Or to ask the ~same question differently: how do we construct a test case?

Reply:
>  if (Py_TYPE(obj) == &PyType_Type) {

This logic will break for metaclasses, where Py_TYPE(obj) is a subclass of PyType_Type, but not PyType_Type itself.

The way you are supposed to do this is: 

> if (PyType_Check(obj)) {

Regardless, it's best to actually not reuse your obj_class_name there because it would be actively misleading in this context. If it's crashing on a PyType object, we will want the inc_ref log to note that it is crashing on a PyType object (and not whatever that PyType object is refering to).

Reply:
> because it would be actively misleading in this context

Agreed. Thanks for pointing out the general flaw. (I think I'll fix that in PyCLIF first, b/o reviewers.)